[2단계 - 장바구니] 기린(정유정) 미션 제출합니다.

jeongyou opened the pull request:
안녕하세요 케빈! 기린입니다🦒
벌써 마지막 미션이라니, 시간이 참 빠르네요...!
마지막 미션을 케빈과 함께할 수 있어서 좋습니다!

이번 리뷰도 잘 부탁드립니다! 
좋은 하루 되세요 😊

## 📦 장바구니 미션

이번 미션을 통해 다음과 같은 학습 경험들을 쌓는 것을 목표로 합니다.

### 1단계

- 클라이언트 상태를 효과적으로 모델링하고 관리할 수 있다.
- Jest, React Testing Library(RTL)를 활용하여 주요 기능에 대한 테스트를 작성할 수 있다.

### 2단계

- 상태를 중복 없이 구성하고, 계산이 가능한 값은 파생 상태로 도출하여 관리할 수 있다.
- 함수(컴포넌트, 훅, util)가 단일한 책임만 갖도록 분리하고, 그 책임에 맞는 명확한 인터페이스를 정의할 수 있다.

## 🕵️ 셀프 리뷰(Self-Review)

### 제출 전 체크 리스트

- [x] 기능 요구 사항을 모두 구현했고, 정상적으로 동작하는지 확인했나요?
- [x] 테스트 케이스를 모두 작성했나요?
- [x] 배포한 데모 페이지에 정상적으로 접근할 수 있나요?
  - 배포 링크 기입: https://jeongyou.github.io/react-shopping-cart/

### 리뷰 요청 & 논의하고 싶은 내용

### 📁 파일 및 폴더 구조 설명
이번 미션에서 OrderReviewPage와 관련된 로직이 모여 있는 order-review 폴더의 구조가 다소 복잡해 보일 수 있어, 간단히 설명드리겠습니다.

order-review는 도메인 단위 폴더이며, 하위 폴더는 각 역할에 따라 다음과 같이 구성되어 있습니다:
- **`hooks/`**
    `OrderReviewPage`에서 사용하는 상태 관련 훅들을 관심사별로 분리해 모았습니다.
    (예: `useShipping`, `useOrderCoupons`, `useOrderInfo` 등)
    
- **`order-coupon/`**
    쿠폰 관련 UI 컴포넌트를 모은 폴더입니다.
    (예: `CouponModal`, `CouponInfo`, `LabelCouponPrice` 등)
    
- **`order-delivery/`**
    배송 지역 관련 UI 컴포넌트를 분리한 폴더입니다.
    (예: 도서산간 지역 선택용 컴포넌트 등)
    
- **`utils/`**
    쿠폰 할인 계산 및 유효성 검사 로직이 포함된 유틸 함수 모음입니다.
    - 쿠폰 타입별 할인 계산 함수는 `utils/calculateCouponDiscount/`에 위치해 있고,
    - `couponDiscountCalculators.ts`에서 이 함수들을 통합해 사용합니다.
    - 쿠폰 조합 계산, 유효성 검사 등은 각각 별도의 파일로 분리해 두었습니다.

---
### 1) 상태 관리 설계 의도
**도서산간 지역 여부 (isJejuOrRemoteArea)**
사용자가 도서산간 버튼을 선택하면 배송비 계산 방식이 달라지기 때문에(+3,000원 추가), useShipping 훅에서 해당 값을 useState로 상태로 관리하였습니다.

**모달 내부 쿠폰 선택 상태 (selectedCouponIds)**
모달 내에서의 쿠폰 선택은 바깥 화면과 동기화되지 않아야 하며, "적용하기" 버튼을 눌렀을 때만 외부 상태가 갱신되어야 합니다.
또한 모달 내부에서 선택한 쿠폰 id에 따라 실시간으로 할인율을 계산해야 했기 때문에, CouponModal 내부에서 selectedCouponIds를 독립적인 상태로 관리했습니다.

**서버에서 받아온 쿠폰 목록 (coupons)**
쿠폰 자동 적용 로직을 실행하기 위해 화면 진입 시 서버로부터 받아온 쿠폰 데이터를 저장해야 했으며, 이를 위해 useCoupon 훅 내부에서 상태로 관리했습니다.

**파생 상태**
위와 같은 핵심 상태들을 기반으로 계산 가능한 값들(적용 가능한 쿠폰, 최적 할인 조합 등)은 모두 파생 상태로 관리하였습니다.
useAvailableCoupons와 useBestCouponCombination 훅 내부에서 파생 상태를 계산하도록 분리하였습니다.

---
### 2) 이번 단계에서 가장 많이 고민했던 문제와 해결 과정에서 배운 점 
### 1. 쿠폰 자동선택
사용자에게 가장 할인 금액이 큰 쿠폰 조합을 추천해주는 UX 적인 측면에서 더 좋다고 생각하여 쿠폰 자동 선택 기능을 추가하였습니다. 

**흐름은 다음과 같습니다. 사용자가 화면에 들어왔을 때**
**- 적용 가능 쿠폰 필터링**
화면 진입 시, 유효기간, 주문 금액, 배송비 여부, 도서산간 선택, 상품 개수 쿠폰의 조건을 바탕으로 사용자가 실제로 적용할 수 있는 쿠폰만 필터링했습니다.

**- 쿠폰 조합 생성**
필터링된 쿠폰들의 id를 이용해, 가능한 모든 쿠폰 조합을 만들었습니다. (최대 2개까지 선택 가능)
※ 쿠폰 적용 순서에 따라 할인 금액이 달라질 수 있기 때문에, 순서가 바뀐 경우도 각각 별도의 조합으로 계산했습니다. (ex: [1,2], [2,1], [2,3], [3,2] 등)

**- 조합별 할인 금액 계산 및 최적 조합 선택**
생성된 모든 조합에 대해 실제 할인 금액을 각각 계산하여, 가장 큰 할인 금액을 제공하는 조합(bestCombination)을 도출했습니다.

**- 자동 선택 및 UI 반영**
사용자가 쿠폰 선택 모달을 열면, 자동으로 최적 조합의 쿠폰이 미리 선택되어 있고, 해당 조합의 할인 금액도 버튼에 표시되도록 구현했습니다.

마지막으로 사용자가 직접 쿠폰을 선택하거나 해제한 이후에는, 자동 선택이 다시 동작하지 않도록 했습니다. 사용자가 의도적으로 쿠폰 선택을 변경했음에도 불구하고 자동으로 다시 쿠폰이 적용되는 것은 사용자 경험 측면에서 오히려 혼란을 줄 수 있다고 판단했습니다!

### 2. 배송비 표시
무료 배송 쿠폰을 적용했을 때, 배송비를 0원으로 표시할지, 아니면 원래 배송비를 그대로 보여줄지에 대해 고민이 있었습니다. 처음에는 무료 배송이 되었으니 배송비도 0원으로 표시하는 것이 맞다고 판단해, 배송비를 0원으로 표기하고 해당 할인 금액을 쿠폰 할인 항목에 포함시켰습니다.

하지만 이런 방식은 사용자 입장에서 혼란을 줄 수 있다는 문제를 발견했습니다.
예를 들어서 다음과 같은 화면이 출력됩니다.
- 주문 금액: 100,000원
- 쿠폰 할인 금액: -3,000원
- 배송비: 0원
- 총 결제 금액: 100,000원

이 경우, 사용자 입장에서는 분명 3,000원이 할인되었는데 총 결제 금액이 왜 그대로인지 의문이 생길 수 있습니다. 실제로는 배송비가 차감된 것이지만, 배송비가 0원으로 표시되다 보니 할인 효과가 사라진 것처럼 보이는 상황이었습니다.

그래서 배송비는 실제 금액인 3,000원을 그대로 표시하고, 쿠폰 할인 금액에만 해당 차감액을 반영하도록 변경했습니다.
최종 화면은 다음과 같이 보이게 됩니다.
- 주문 금액: 100,000원
- 쿠폰 할인 금액: -3,000원
- 배송비: 3,000원
- 총 결제 금액: 100,000원

이렇게 구성하면, 사용자가 어디에서 할인되었는지 더 명확하게 이해할 수 있고, 전체적으로 더 일관된 UX를 제공할 수 있다고 판단했습니다.

---
### 3) 이번 리뷰를 통해 논의하고 싶은 부분
### 1. 상태 관리 방식
쿠폰 자동 적용 로직을 위해, 페이지 진입 시점에 totalDiscount와 selectedCouponIds 등의 상태를 설정하고 화면에서 사용하고 있습니다. 하지만 사용자가 쿠폰 선택 모달에 진입한 이후에는 쿠폰을 새롭게 선택할 수 있기 때문에, 모달 내부에서도 별도의 선택 상태를 관리해야 하는 상황입니다.

또한 모달 내부에서 선택한 쿠폰에 따라 실시간으로 할인 금액(totalDiscount)도 재계산되어야 하므로, 초기에 설정된 selectedCouponIds와 totalDiscount는 사실상 초기 진입 시점에만 사용되고 이후로는 계속 새로운 값으로 덮어쓰게 됩니다.

그리고 모달을 닫고 선택한 쿠폰을 적용할 때, 화면 외부의 상태로 다시 반영하면서 한 번 더 할인 금액 계산이 이루어지기 때문에
**초기 상태 → 모달 내부 상태 → 최종 적용** 흐름 속에서 상태가 여러 군데로 나뉘어 관리되고 있습니다.

동기화 자체는 문제 없이 잘 이루어지고 있지만, 상태가 지나치게 분산되고, 일시적으로만 쓰이는 상태들도 많아지다 보니 전체 구조가 복잡해졌다는 느낌을 받고 있습니다. 하나의 상태가 틀어지거나 잘못된 값이 반영되는 경우는 없지만, 유지보수나 로직 이해 측면에서 부담이 커진 것 같습니다.

❓이런 상황에서 궁금한 점은 두가지인데요!
- 초기 자동 적용에만 사용되는 상태와, 이후 사용자 선택에 따라 계속 활용되는 상태를 어떻게 명확하게 구분하고 관리하면 좋을지
- SSOT(Single Source of Truth) 원칙을 유지하면서도 자동 적용과 사용자 선택을 유연하게 함께 다룰 수 있는 구조는 어떤 형태가 적절할지

케빈과 함께 논의해보고 싶습니다!

### 2. 페이지 단위 훅에 대한 고민
현재 OrderReviewPage에서는 useOrderInfo, useShipping, useOrderCoupons, useModal, useNavigation 등 관심사 별로 분리된 여러 훅들을 개별적으로 호출하고 있습니다.
처음에는 이 훅들을 useOrderReviewPage라는 하나의 훅으로 감싸서 통합적으로 반환할까 고민했지만, 그렇게 할 경우 리턴해야 하는 값이 20개 가까이 되어버려 실제로는 그렇게 구현하지 않았습니다.

그렇지만 페이지 내에서 훅이 여러 개 사용되다 보니 코드가 다소 산만해 보인다는 느낌도 있고 어차피 해당 페이지에서만 사용하는 훅들이라면 페이지 단위의 훅 하나로 묶고, 내부에서 관심사 별 훅을 호출한 뒤 필요한 값만 반환하는 구조가 오히려 더 가독성이 좋고 관리하기 쉬운 방향일 수도 있지 않을까? 하는 생각도 들었습니다.

❓이런 경우, 하나의 페이지 훅으로 묶는 편이 더 나은 선택인지, 아니면 현재처럼 관심사별 훅들을 페이지에서 개별적으로 사용하는 것이 더 명확한 구조인지에 대해 케빈의 의견이 궁금합니다!

---
## ✅ 리뷰어 체크 포인트

<!-- 리뷰어가 이 PR을 검토할 때 중점적으로 확인할 사항입니다.
코드의 완성도뿐만 아니라, 리뷰이가 구현 과정에서 어떤 고민과 결정을 하며 학습했는지도 함께 고려해 주세요. -->

### 1. 상태관리

- 원본상태/파생상태를 적절히 구분하여 선언하였나요?
- React state 를 불필요하게 선언한 부분은 없나요?
- 상태 관리 로직의 책임이 적절히 응집/분리되었나요? (ex. reducer, hook, Context)

### 2. 역할과 책임

- 단일 책임 원칙(SRP)에 맞게 컴포넌트/함수가 분리되었나요?
- 컴포넌트/함수의 인터페이스가 구현에 의존하지 않고, 책임에 맞게 설계되었나요?

### 3. 테스트

- 쿠폰 기능에 대한 다양한 시나리오가 충분히 테스트되었나요?
- 함수의 책임을 분리하고, 순수성을 유지하는 등 테스트하기 좋은 코드를 작성하였나요?


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Introduced an order review page with coupon selection modal, delivery region toggle, and detailed payment summary.
  - Added support for applying up to two simultaneous coupons with automatic best discount selection.
  - Implemented multiple coupon types: fixed discount, BOGO, free shipping, and time-limited percentage discount.
  - Enhanced shipping fee policy with regional surcharges and free shipping thresholds.
  - Added modal components, styled buttons, and informational notices for improved user interaction.
  - Persisted cart item selection state in localStorage for consistent user experience.
  - Added detailed order summary and coupon info components for clearer checkout flow.

- **Bug Fixes**
  - Improved cart item selection persistence and pricing calculations in the shopping cart.

- **Documentation**
  - Updated and expanded README files with detailed feature breakdowns and UI component descriptions.

- **Tests**
  - Added comprehensive tests for coupon logic, modal behavior, and order review page rendering.

- **Chores**
  - Updated dependencies, switched to Emotion for styling, and adjusted TypeScript and Vite configurations for improved development experience.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

jeongyou committed:
docs: step1 기능 구현 목록

jeongyou committed:
docs: step2 기능 구현 목록 작성

jeongyou committed:
chore: react-swc 플러그인을 react 플러그인으로 교체하고 Emotion 설정 추가

jeongyou committed:
feat: 모달 컴포넌트

jeongyou committed:
feat: BaseButton 컴포넌트 구현

jeongyou committed:
fix: handleCheckout 핸들러의 네비게이션 경로를 /order-confirm → /order-review 로 수정

jeongyou committed:
feat: 주문서 확인 페이지(/order-review) 라우트 추가

jeongyou committed:
feat: 주문서 확인용 주문된 장바구니 아이템 리스트 컴포넌트 추가

jeongyou committed:
feat: 주문서 결제 금액 요약 컴포넌트(LabelCouponPrice) 추가

jeongyou committed:
feat: 쿠폰 테이터 요청 api 함수 작성

jeongyou committed:
feat: api 요청으로 온 쿠폰 데이터 타입 정의

jeongyou committed:
style: Modal이 center 포지션일 때 x, y 축모두 정중앙에 위치하도록 transform: translate(-50%, -50%);로 수정

jeongyou committed:
feat: 로딩 상태, 에러 핸들링을 관리하는 API 요청 커스텀 훅 구현

jeongyou committed:
feat: 쿠폰 테이터 fetching을 관리하는 useCoupon 커스텀 훅 구현

jeongyou committed:
style: 페이지의 위, 아래 padding 0으로 수정

jeongyou committed:
feat: Info 메세지를 보여주는 InfoNotice 컴포넌트 구현

jeongyou committed:
style: 불필요한 padding and margin을 OrderItemList 컴포넌트에서 제거

jeongyou committed:
feat: CouponModalOpenButton 컴포넌트 추가

jeongyou committed:
chore: 줄바꿈 추가

jeongyou committed:
refactor: 체크박스 컴포넌트 셀프 클로징태그로 변경, ItemList 높이 조정

jeongyou committed:
refactor: InfoBox 직접 구현 대신 InfoNotice 컴포넌트로 대체하여 코드 단순화

jeongyou committed:
feat: ModalBody 컴포넌트에 height prop 추가 및 스타일 수정

jeongyou committed:
refactor: ItemList 컴포넌트의 높이를 380px에서 260px로 수정

jeongyou committed:
feat: 쿠폰 정보 표시 컴포넌트 CouponInfo 구현

jeongyou committed:
feat: 배송 지역 안내 메시지와 체크박스 결합한 DeliveryRegionCheckbox 컴포넌트 구현

jeongyou committed:
feat: 주문 확인 페이지 상단 정보(OrderPageInfo) 컴포넌트 분리 및 스타일 적용

jeongyou committed:
feat: 쿠폰 선택 모달 컴포넌트(CouponModal) 구현

jeongyou committed:
feat: 배송 정보 영역 컴포넌트 구현

jeongyou committed:
feat: 'HH:mm:ss' 형식의 시간을 '오전/오후 N시'로 변환하는 formatKoreanTime 함수 추가

jeongyou committed:
feat: CouponInfo 컴포넌트에 한글 날짜·금액·시간 포맷 적용 및 스타일링 리팩터링

jeongyou committed:
feat: 주문 확인 페이지 구현

jeongyou committed:
feat: 장바구니 context에서 선택 상품 및 주문/배송/총 결제 금액 계산값 제공

jeongyou committed:
refactor: OrderListContext의 selectedItems, orderPrice, shippingFee, totalPrice 등 계산값을 사용하도록 관련 컴포넌트 리팩터링

jeongyou committed:
feat: useCoupon 훅에 fetchCoupons 호출을 useEffect로 추가하여 초기 데이터 로딩 구현

jeongyou committed:
feat: CouponInfo에 disabled prop 추가, disabled 스타일링 추가

jeongyou committed:
feat: CouponModal에서 사용 가능한 쿠폰에 따라 CouponInfo의 disabled prop 설정

jeongyou committed:
feat: 적용 가능한 쿠폰을 조회하는 useAvailableCoupons 훅 추가

jeongyou committed:
feat: 가장 높은 할인율의 쿠폰 조합을 찾는 useBestCouponCombination 훅 추가

jeongyou committed:
feat: 가능한 모든 쿠폰 조합을 생성하는 useCouponCombinations 훅 추가

jeongyou committed:
refactor: 쿠폰 관련 로직을 분리된 훅들로 조합하는 useApplyCoupon 훅 구현

jeongyou committed:
feat: 가장 비싼 상품을 찾는 getMostExpensiveItem 유틸 함수 추가

jeongyou committed:
feat: BOGO 쿠폰 적용 가능한 상품을 필터링하는 getBogoEligibleItems 함수 추가

jeongyou committed:
feat: BOGO 쿠폰 할인 금액을 계산하는 calculateBogoDiscount 함수 추가

jeongyou committed:
feat: 미라클세일 쿠폰 할인 금액을 계산하는 calculateMiracleSaleDiscount 함수 추가

jeongyou committed:
feat: FREESHIPPING 쿠폰 할인 금액을 계산하는 calculateFreeShippingDiscount 함수 추가

jeongyou committed:
feat: FIXED5000 쿠폰 할인 금액을 계산하는 calculateFixed5000Discount 함수 추가

jeongyou committed:
feat: 쿠폰 유효기간을 검증하는 isCouponValid 함수 추가

jeongyou committed:
feat: 쿠폰별 적용 가능 여부를 검증하는 couponValidator 함수 추가

jeongyou committed:
feat: 쿠폰별 할인 계산 함수를 매핑하는 couponDiscountCalculators 객체 추가

jeongyou committed:
feat: 쿠폰별 할인 금액을 계산하는 getDiscountByCouponId 함수 추가

jeongyou committed:
feat: 가능한 모든 쿠폰 조합 ID를 생성하는 getAllCouponCombinationIds 함수 추가

jeongyou committed:
feat: 쿠폰 모달 표시/숨김을 관리하는 useModal 훅 추가

jeongyou committed:
feat: 주문 상품 개수와 상태를 관리하는 useOrderInfo 훅 추가

jeongyou committed:
feat: 배송비 계산과 지역 설정을 관리하는 useShipping 훅 추가

jeongyou committed:
feat: 주문 페이지 네비게이션을 관리하는 useNavigation 훅 추가

jeongyou committed:
refactor: 주문 검토 페이지 로직을 통합하는 useOrderReview 훅 추가

jeongyou committed:
feat: 쿠폰 컴포넌트에 자동 선택 기능 추가

jeongyou committed:
feat: 쿠폰 모달에 최적 쿠폰 자동 선택 및 할인 금액 표시 추가

jeongyou committed:
feat: 쿠폰 할인 및 배송비 동적 계산 기능 추가

jeongyou committed:
feat: 주문 검토 페이지에 쿠폰 조합 및 최적 쿠폰 자동 선택 기능 추가

jeongyou committed:
refactor: best coupon combination 로직에 useMemo 사용

jeongyou committed:
feat: getCouponCombinations 추가

jeongyou committed:
refactor: useMemo를 사용하여 쿠폰 유효성 검사 최적화

jeongyou committed:
refactor: 쿠폰 조합 계산 로직 분리 및 코드 구조 개선

jeongyou committed:
refactor: Header 컴포넌트 left props 옵셔널로 변경

jeongyou committed:
refactor: 불필요한 인자 제거

jeongyou committed:
refactor: 결제 확인 페이지 구조 개선 및 불필요한 컴포넌트 제거

jeongyou committed:
feat: 쿠폰 선택을 담당하는useOrderCoupons 훅 구현

jeongyou committed:
refactor: 불필요한 커스텀 훅, 함수 삭제

jeongyou committed:
feat: '/order-confirm'  로 이동할 때 state로 payment값 전달하도록 수정

jeongyou committed:
refactor: payment값을 내부에서 계산하지 않고 외부에서 받아온 값을 보여주도록 수정

jeongyou committed:
feat: CouponModal에 쿠폰 선택 기능 개선 및 내부에서만 사용할 modalDiscount 계산 로직 추가

jeongyou committed:
refactor: CouponInfo 컴포넌트에서 체크 상태 관리 로직을 부모 컴포넌트로 이동

jeongyou committed:
refactor: 주문 검토 페이지 로직 분리 및 실시간 할인 금액 적용

- useOrderCoupons 훅으로 쿠폰 관련 로직 통합
- 실시간 선택된 쿠폰 할인 금액 계산 및 표시
- 최종 결제 금액을 다음 페이지로 전달
- 불필요한 상태 관리 코드 제거

jeongyou committed:
refactor: bestCouponIds prop 제거, useOrderCoupons hook 에서 가져다 쓰도록 변경

jeongyou committed:
refactor: selectedCouponIds 상태를 bestCouponIds로 초기화

jeongyou committed:
refactor: 불필요한 props 제거

jeongyou committed:
refactor: 테스트 타입을 jest에서 vitest로 변경

jeongyou committed:
feat: couponMockData 쿠폰 목데이터 추가

jeongyou committed:
refactor: discount를 옵셔널로 변경

jeongyou committed:
refactor: GoToShoppingCart의 text를 '장바구니로 돌아가기'로 변경

jeongyou committed:
refactor: 테스트를 위한 data-testid 추가

jeongyou committed:
test: 쿠폰 비활성화 테스트 추가

jeongyou committed:
test: 쿠폰 모달에 대한 유닛 테스트 추가

jeongyou committed:
test: OrderReviewPage 통합 테스트 추가

jeongyou committed:
test: useAvailableCoupons 훅에 대한 유닛 테스트 추가

jeongyou committed:
test: useBestCouponCombination 훅에 대한 유닛테스트 추가

jeongyou committed:
feat: 장바구니 선택 상태를 로컬스토리지에 저장하는 기능 추가

JeongBin0227 commented on src/components/common/Modal/Modal.tsx:
```diff
@@ -0,0 +1,117 @@
+/** @jsxImportSource @emotion/react */
```
질문) 이건 무슨 용도예요?

jeongyou replied:
처음에 emotion 설정을 잘못하여서 css props 인식이 안되어서 임시 방편으로 Emotion의 css props를 자동으로 JSX에서 인식하게 하게 하기 위해서 이 pragma를 사용하였는데, 그 역할을 해주는  @emotion/react/types/css-prop를 types tsconfig.json의 types에 넣어서 이 주석을 사용하지 않아도 되어서 제거하였습니다!

관련 커밋: 
- [feat: @emotion/react/types/css-prop를 types tsconfig.json의 types에 넣음](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/b0e48e9ceb1b0138f8051e116a2d6f0b16247383)

- [refactor: 불필요한 JSX pragma 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/9b844f7b2b86e87a9542a4bfeb9125143f606d67)

JeongBin0227 commented on src/components/common/Modal/Modal.tsx:
```diff
@@ -0,0 +1,117 @@
+Modal.Title = ModalTitle;
+Modal.Button = BaseButton;
+Modal.ConfirmButton = ConfirmButton;
+Modal.CancelButton = CancelButton;
```
코드는 전체적으로 괜찮은데 구조측면에서
```
 Modal
 ┣ Modal.tsx // Provider 및 Wrapper 정의
 ┣ ModalHeader.tsx
 ┣ ModalBody.tsx
 ┣ ModalFooter.tsx
 ┣ ModalTitle.tsx
 ┣ ModalContext.ts
 ┗ styles.ts
```
이렇게 가져갔어도 Modal.tsx가 너무 많은 책임을 지지 않도록 역할 분산할수 있을거 같아요

jeongyou replied:
해당 모달 컴포넌트는 레벨2 페이먼츠 모듈에서 만들었던 모달을 가져와 재사용한 것인데, 한 파일에 모두 작성하는 것보다 파일별로 역할에 따라 구조를 분리하는 방식이 훨씬 더 명확하게 책임을 드러낼 수 있는 것 같아서 이 점을 반영하여 구조를 수정하였습니다!

관련 커밋: [refactor: 모달 컴포넌트를 역할에 따라 파일별로 분리](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/4c54686dcd150bfa85e889f4f46f1b3cf7ba0ef1)

JeongBin0227 commented on src/components/common/Modal/styles/Button.styles.ts:
```diff
@@ -0,0 +1,61 @@
+  cursor: ${disabled ? 'not-allowed' : 'pointer'};
+`;
+
+export const CancelButtonStyle = (
```
이렇게 특정 조건을 가지고 개발할때는 객체 매핑도 많이 사용됩니다

```
// 개선안
const variationStyles = {
  primary: css`
    background-color: black;
    color: white;
    border: none;
    &:hover {
      background-color: #222;
    }
  `,
  secondary: css`
    background-color: white;
    color: #333;
    border: 1px solid #ccc;
    &:hover {
      background-color: #f5f5f5;
    }
  `,
};

export const CancelButtonStyle = (
  variation: 'primary' | 'secondary' = 'secondary',
  widthSize: string
) => css`
  padding: 0 18px;
  height: 36px;
  border-radius: 5px;
  cursor: pointer;
  width: ${widthSize};
  ${variationStyles[variation]};
`;
```
이렇게 사용됩니다.

jeongyou replied:
조건문으로 스타일 분기를 하다 보니 코드가 길어지고 가독성도 떨어지는 느낌이 있었는데, 리뷰 주신 객체 매핑 방식이 훨씬 간결하고 명확해서 해당 방식으로 반영했습니다!
좋은 방향 제안해주셔서 감사합니다. 앞으로는 스타일 분기 시 객체 매핑 방식을 사용하는게 더 좋을 것 같네요!

관련 커밋: [refactor: CancelButton 스타일 조건 분기를 객체 매핑 방식으로 리팩토링](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/8e63cdaf1206c8097d7346db88c47f46e968117b)

JeongBin0227 commented on src/components/hooks/useFocusTrap.ts:
```diff
@@ -0,0 +1,55 @@
+      }
+    };
+
+    container.addEventListener("keydown", handleKeyDown);
```
질문) container 에 등록한 이유가 있을까요?  포커스는 container 바깥에 있을 때도 있을거 같아서 document 에 거는게 안전한가..? 생각이 들어서요

jeongyou replied:
앗 이 코드는 페이먼츠 모듈 미션의 모달에서 사용하기 위해 만든 훅인데요, 모달이 활성화된 동안 포커스가 container 내부에 있다고 가정하고, 내부 요소들 사이에서만 포커스를 순환시키는 목적으로 container에 이벤트를 등록했습니다.

제가 이 훅을 기존 모달 코드에서 재사용한 부분이라는 걸 미리 말씀드렸어야 했는데, 전달을 못 드려서 죄송합니다...! 🙇‍♀️

JeongBin0227 commented on src/components/hooks/useFocusTrap.ts:
```diff
@@ -0,0 +1,55 @@
+    return () => {
+      container.removeEventListener("keydown", handleKeyDown);
+    };
+  }, [active]);
```
만약 필요하다면 containerRef.current를 의존에 넣어도 되겠네요
ref는 DOM 변화와 동기화되지 않으니까요!

jeongyou replied:
useEffect 내부에서 containerRef가 사용되었지만, 말씀해주신 것처럼 ref는 항상 동일한 객체를 유지하기 때문에 의존성 배열에 넣더라도 실제 변화로 감지되지 않고 리렌더링도 발생하지 않는 것으로 알고 있습니다! 그래서 현재 상황에서는 굳이 의존성 배열에 포함하지 않아도 괜찮을 것 같습니다!

❓그런데 혹시 말씀해주신 "필요한 경우"라는 게 ESLint 경고가 뜨는 상황을 말씀하신 걸까요?
제가 코멘트의 의도를 잘못 이해하고 엉뚱한 방향으로 답변드린 건 아닐까 걱정되어서 여쭤봅니다...!

JeongBin0227 replied:
네 제가 말씀드린 필요한 경우는 기린이 말한것처럼 ESLint의 exhaustive-deps 경고가 뜨거나, 혹은 ref.current가 나중에 설정되는 경우(첫 마운트 시점 이후 DOM 변화로 접근이 필요한 경우 등등..)를 의미했습니다. 현재 코드처럼 ref의 초기 접근 정도만 필요한 상황이라면 오히려 의존성 배열에 포함하지 않는 쪽이 명확하고 불필요한 리렌더링도 방지할 수 있어 더 적절한 선택이라 생각합니다 😊~

JeongBin0227 commented on src/components/hooks/useKeyPress.ts:
```diff
@@ -0,0 +1,26 @@
+
+    window.addEventListener(eventType, keyListener);
+    return () => window.removeEventListener(eventType, keyListener);
+  }, [key, handler, eventType]);
```
의문) handler가 의존성에 직접 들어가 있는데이러면 useCallback 없이 사용되면 handler가 변경되면 매번 리스너 등록/해제 발생할거 같아요

jeongyou replied:
말씀해주신 부분 확인해보니까 정말 handler가 useCallback 변경되면 useEffect가 다시 실행되겠네요...!
그래서 코드에서 handler로 념겨주는 함수를 useCallback으로 감싸서 동일한 함수가 유지되도록 처리했습니다!

관련 커밋: [refactor: useCallback을 사용하여 handleShowCouponModal 최적화](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/1b46511f2a68a7ac0894ef4d70babc58115f21f1)

그리고 더 찾아보니 handler가 자주 바뀔 수 있지만 리스너 등록은 유지하고 싶을 때는 아래처럼 useRef를 활용해서 핸들러를 추적하는 방식도 있더라고요!

```ts
const handlerRef = useRef(handler);

useEffect(() => {
  handlerRef.current = handler;
}, [handler]);
```

이 경우는 이벤트 리스너는 한 번만 등록하고 내부 로직만 최신 상태를 참조할 수 있어서 상황에 따라 고려해볼 수도 있을 것 같아요!

JeongBin0227 commented on src/pages/order-review/OrderPageInfo.tsx:
```diff
@@ -0,0 +1,43 @@
+  <PageInfo>
+    <InfoTitle>주문 확인</InfoTitle>
+    <Description
+      aria-label={`총 ${typeCount}종류의 상품 ${totalCount}개를 주문합니다.`}
```
자세하네요 👍

jeongyou replied:
감사합니다!

JeongBin0227 commented on src/pages/order-review/OrderReviewPage.tsx:
```diff
@@ -0,0 +1,131 @@
+  } = useOrderCoupons(isJejuOrRemoteArea);
+
+  const { selectedItems, orderPrice } = useOrderListContext();
+  const currentDiscount = useMemo(() => {
+    const selectedCoupons = selectedCouponIds
+      .map((id) => availableCoupons.find((c) => c.id === id))
+      .filter(Boolean) as CouponContent[];
+
+    return selectedCoupons.reduce((total, coupon) => {
+      return (
+        total +
+        getDiscountByCouponId(
+          coupon,
+          orderPrice,
+          selectedItems,
+          isJejuOrRemoteArea
+        )
+      );
+    }, 0);
+  }, [
+    selectedCouponIds,
+    availableCoupons,
+    orderPrice,
+    selectedItems,
+    isJejuOrRemoteArea,
+  ]);
+
+  const payment = orderPrice + actualShippingFee - currentDiscount;
```
currentDiscount와 payment 계산이 컴포넌트 내부에 있는데, useOrderInfo 또는 새로운 usePaymentInfo 훅으로 파생 상태도 캡슐화할수 있을거 같아요

jeongyou replied:
useCalculateTotalDiscount와 usePayment훅을 만들어서 currentDiscount와 payment의 계산 로직을 캡슐화 하였습니다!

관련 커밋:
- [feat: 선택된 쿠폰들의 할인 금액을 구하는 useCalculateTotalDiscount 훅 구현](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/2dff3e50f2a901675caa0e198cef3ed58a390830)
-> 처음에는 훅으로 구현했는데 내부 로직이 커스텀 훅의 조건에 해당하지 않아서 단순 함수로 변경하였습니다. 
-> [feat: 선택된 쿠폰들의 할인 금액을 구하는 calculateTotalDiscount 함수 구현](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/a2ff77e29abecc0d45fd72fa4f0d031ec74c0caa)
- [feat: 결제 금액을 계산하는 usePayment 훅을 구현](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/07268f3fb1130b1d6ab0e0440b356c94795643f2) 
- [refactor: 페이지에서 직접 currentDiscount, payment를 계산하지 않고 커스텀 훅을 사용하여 계산하도…](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/3f7f24977b32fc3e4bcc8fef143b799dc35c5416)

JeongBin0227 commented on src/pages/order-review/OrderReviewPage.tsx:
```diff
@@ -0,0 +1,131 @@
+          coupons={coupons ?? []}
+          isLoading={isLoading}
+          availableCoupons={availableCoupons}
+          // selectedCouponIds={selectedCouponIds}
```
주석은 필요없다면 지워도 될거같아요!

jeongyou replied:
불필요한 주석을 삭제하였습니다!

관련 커밋: [refactor: 페이지에서 직접 currentDiscount, payment를 계산하지 않고 커스텀 훅을 사용하여 계산하도…](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/3f7f24977b32fc3e4bcc8fef143b799dc35c5416)
-> 이 커밋에 같이 올라갔네요😅

JeongBin0227 commented on src/pages/order-review/__test__/CouponDisabled.test.tsx:
```diff
@@ -0,0 +1,298 @@
+      }
+    }
+  );
+});
```
지금 페이지 단위로 나눠져있는데 페이지 단위보단 도메인 단위나 목적별로 해도 좋을거같아요

JeongBin0227 replied:
각 테스트에서 mockReturnValue, render, fireEvent, waitFor, getByTestId 등 중복되는데 중복을 줄여봐도 좋을거 같아요

JeongBin0227 replied:
혹시 여기는 직접 짜신 코드일까요..? 뭔가 커서의 향기가 나서요..

jeongyou replied:
> 지금 페이지 단위로 나눠져있는데 페이지 단위보단 도메인 단위나 목적별로 해도 좋을거같아요

처음에는 통합 테스트를 중심으로 작성할 생각이어서 페이지 단위로 구성하게 되었는데, 돌이켜보면 이번 미션에서는 쿠폰 관련 훅을 중심으로 테스트를 작성하는 게 더 적절했을 것 같습니다. 피드백 참고해서 도메인이나 목적별로 구조를 나눠보겠습니다!

---
> 각 테스트에서 mockReturnValue, render, fireEvent, waitFor, getByTestId 등 중복되는데 중복을 줄여봐도 좋을거 같아요

넵! 중복되는 부분은 유틸 함수나 공통 setup 함수로 추출해서 개선해보겠습니다!

---
> 혹시 여기는 직접 짜신 코드일까요..? 뭔가 커서의 향기가 나서요..

사실 경계값 케이스를 빠르게 정리하고 싶어서 GPT의 도움을 받아 초안을 만들고, 그걸 기반으로 mock 설정이나 흐름은 제가 직접 확인하면서 수정해 테스트 코드를 작성했습니다..!
아직 테스트 작성이 익숙하지 않은 데다 시간이 부족하다 보니, 빠르게 처리해야겠다는 마음에 AI의 도움을 받았습니다!
테스트 코드에 대해서는 앞으로 좀 더 직접 고민하고, 가능한 한 스스로 작성해보도록 하겠습니다.

jeongyou replied:
주요 로직인 useAvailableCoupons와 useBestCouponCombination 훅의 테스트 코드를 우선 정리하고, 나머지 테스트에서도 중복되는 mock 설정 및 렌더링 로직은 공통 setup 함수로 분리해 중복을 줄였습니다.

관련 커밋:
- [test: useBestCouponCombination 테스트 중복 로직 헬퍼 함수로 리팩토링](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/fe3c3ae53a15473cb56954ca59851bc64bf2de1c)
- [test: useAvailableCoupons 테스트 헬퍼 함수로 리팩토링](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/6bae8f49dcc840329951f94c634a59d027a153d8)
- [test: OrderReviewPage 쿠폰 모달 테스트 중복제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/ecaa0e1bb680def72472ee0e741deea279afa0c0)
- [test: CouponDisabled 테스트 유틸 함수 추출 및 중복 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/84c0d298c0be8ddbf9a1f3a0db30e738055615bb)

현재는 useAvailableCoupons, useBestCouponCombination이 두 훅이 핵심 비즈니스 로직이라고 판단하여 해당 부분만 테스트했는데, 케빈이 보시기에 추가적인 테스트가 필요하다고 생각되는 부분이 있을까요..? 혹시 제가 놓친 테스트 포인트가 있다면 피드백 부탁드립니다...! 

그리고 혹시라도 제가 리뷰 내용을 잘못 이해하고 있는 부분이 있다면 말씀해 주시면 감사하겠습니다!

JeongBin0227 replied:
쭉 확인해봤는데 제가 생각한 주요 포인트는 다 작성된거 같아요

JeongBin0227 commented on src/pages/order-review/hooks/useAvailableCoupons.ts:
```diff
@@ -0,0 +1,22 @@
+  coupons: CouponContent[],
+  isJejuOrRemoteArea: boolean
+) => {
+  const { selectedItems, orderPrice } = useOrderListContext();
+
+  const availableCoupons = useMemo(() => {
+    return couponValidator(
+      coupons,
+      isJejuOrRemoteArea,
+      selectedItems,
+      orderPrice
+    );
+  }, [coupons, isJejuOrRemoteArea, selectedItems, orderPrice]);
```
지금 useMemo 를 쓰고 계시니까 
const { selectedItems, orderPrice } = useOrderListContext();

하기 보단
const orderContext = useOrderListContext();

이렇게 해도 좋을거 같아요

왜냐하면 useMemo 의존성 배열에서 selectedItems, orderPrice를 직접 넣으면 참조 변경에 의한 재계산 우려 있는데 Context 객체 자체를 의존성에 넣고 내부 필드는 couponValidator에서 접근하게 하거나, shallowEqual 검토하면 더 좋을거 같아요

jeongyou replied:
const orderContext = useOrderListContext();처럼 context 전체를 묶어서 사용하는 방법을 생각해보았는데,  useOrderListContext() 내부에서 매번 새로운 객체를 리턴하고 있기 때문에, orderContext로 묶어도 selectedItems, orderPrice를 따로 사용하는 것과 실질적으로 차이가 없다고 판단했습니다!

그리고 다시 보니까 couponValidator 함수 자체가 복잡하거나 무거운 계산을 수행하는 것이 아니라서, useMemo로 메모이제이션해도 큰 이점이 없다고 생각했습니다
오히려 불필요하게 의존성을 관리해야 하므로, 해당 useMemo는 제거하는 것이 더 나은 방향이라고 판단하여 제거하였습니다. 

❓혹시 이 부분에 대해 케빈은 어떻게 생각하시나요? 제가 케빈의 의도를 잘못 이해한 부분이 있을 수도 있어 의견 주시면 감사하겠습니다!

관련 커밋: [refactor: 불필요한 useMemo 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/0a8682ca12de7c15bddc669bc01ed03146fd1d52)

JeongBin0227 replied:
흠.. 기린이 적어준 판단과 접근이 너무 좋은거 같아요! 저도 합리적이라고 생각합니다

제가 의도했던 개선 방향은, useMemo를 굳이 유지한다면 불필요한 디스트럭처링으로 인한 의존성 관리 복잡성을 줄이기 위한 것이었는데, 기린이 말해준것처럼 useOrderListContext() 내부가 매 호출마다 새로운 객체를 반환한다는 점과 결국 객체로 묶든, 필드를 디스트럭처링하든 의존성 배열 측면에서는 동일한 재계산이 발생하니, 이 구조에서는 큰 의미가 없다는 점에 공감되는거 같아요
또한 couponValidator가 무거운 연산을 수행하지 않는다는 점까지 고려해서 useMemo 자체를 제거한 판단도, 
불필요한 메모이제이션보다는 단순성과 명확성을 택한 점이 유지보수에도 훨씬 이점이 있다는것에 동의하고 합리적인거 같아요! 좋은 의견 감사해요~

JeongBin0227 commented:
코드 중간중간 최적화 (useMemo 등등..) 을 쓰신곳이 있고 아닌곳이 있는데 기준이 궁금합니다!

JeongBin0227 commented:
태스트 코드에 대해서 궁금한데요! 뭔가 정리가 잘되있는거 같으면서도 중구난방한 느낌이 많아서 어떤 기준으로 짜시는지 궁금해요

JeongBin0227 commented:
주석을 자주 쓰시는데 어떤 방향성을 가지고 쓰시는지도 궁금해요

JeongBin0227 commented on src/pages/order-review/hooks/useOrderCoupons.ts:
```diff
@@ -0,0 +1,73 @@
+    setAllCouponCombinationIds(combinations);
+  };
+
+  const handleSelectCoupons = (newCoupons: number[]) => {
```
지금 2가지 책임을 가지는거 같은데 (업데이트, 재생성) 로직을 분리해봐도 좋을거 같아요

jeongyou replied:
이 리뷰를 반영하기 전에 useOrderCoupon에서 allCouponCombinationIds가 불필요하다는 것을 깨달았습니다...! 

availableCoupons를 기반으로 최적의 쿠폰 조합 (bestCouponIds) 을 계산하고 있었는데  또 한 번 getAllCouponCombinationIds() 함수를 사용해서 모든 쿠폰 조합을 계산해서 allCouponCombinationIds라는 상태에 따로 저장하고 있었습니다. 

결론적으로 이 훅 안에서 쿠폰 조합을 두 번 계산하고 있었고,  bestCouponIds만으로 이미 필요한 로직이 다 되니까, 굳이 allCouponCombinationIds를 따로 저장하고 관리할 필요가 없다는 것을 깨달았습니다. 그래서 allCouponCombinationIds와 관련된 코드 먼저 제거 하는 작업을 진행하였습니다!

관련 커밋: [refactor: useOrderCoupons 훅에서 allCouponCombinationIds 상태 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/54889964c263b1cdcb984bc9248984f2c0cb6f4c)

그리고 정리된 훅의 나머지 코드에서도 선택된 쿠폰 id들의 상태 업데이트와 재생성 로직 두가지 책임을 가지고 있어서 로직을 분리하여 useCouponSelection이라는 훅을 구현하였습니다!

관련 커밋: 
- [feat: 선택된 쿠폰의 상태를 관리하는 useCouponSelection 훅 구현하여 useOrderCoupons에서 로직 분리](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/7512d604798c2d181b85d0f6607773bfdc7e5bb9)
- [refactor: 페이지에서 useCouponSelection 사용하도록 수정](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/4f6933134dcf51729d06a35b5b1c8cd88bdf270e)

JeongBin0227 commented on src/pages/order-review/hooks/useOrderCoupons.ts:
```diff
@@ -0,0 +1,73 @@
+  );
+
+  const [selectedCouponIds, setSelectedCouponIds] =
+    useState<number[]>(bestCouponIds); //이거 지울 수 있음
```
지우면 이슈가 생기지 않나요??

jeongyou replied:
헉, 이 부분의 주석이 지워지지 않았네요...!
처음에는 bestCouponIds를 모달에 바로 뿌려주기만 해도 쿠폰을 자동 선택할 수 있다고 생각해서 selectedCouponIds가 필요 없다고 판단했습니다.

그런데 다시 생각해보니까 자동 추천된 bestCouponIds는 초기값일 뿐이고 selectedCouponIds를 새로 set 하는 핸들러가 모달에 넘겨져서, 사용자가 모달 안에서 직접 쿠폰을 선택한 후 적용하기 버튼을 눌렀을 때 변경되어야 하는 값이기 때문에 이 상태를 따로 관리해주는 게 맞다는 걸 깨달았습니다.

그리고 적용하기 버튼을 누른 후 모달 밖의 페이지에서는 사용자가 선택한 쿠폰의 id(selectedCouponIds)를 기반으로 할인 금액 등을 다시 계산해야 하므로, 이 상태는 꼭 유지되어야 하는 것이였습니다!

제가 불필요한 주석을 남겨서 오히려 혼란을 드린 것 같아 죄송합니다... 😥

JeongBin0227 commented on src/pages/order-review/hooks/useOrderCoupons.ts:
```diff
@@ -0,0 +1,73 @@
+    // 핸들러
+    handleSelectCoupons,
+  };
+};
```
hook의 반환 데이터가 많아서 반환값을 타입으로 정리해도 좋을거 같아요

jeongyou replied:
코드 리뷰를 통해 이 훅을 리팩터링 하면서, 불필요한 상태와 로직을 제거하고 역할에 따라 훅을 분리하면서 반환값이 5개로 줄어들었습니다. 덕분에 현재는 타입 추론이 크게 어렵지 않아 따로 타입을 분리하진 않았습니다.

최종 정리된 훅의 형태는 아래 커밋에서 확인하실 수 있습니다!
관련 커밋: [style: 불필요한 줄바꿈 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/11c69b071f8049c0306b14761bbb31c5c7236906)

아직 타입스크립트에 익숙하지 않아서 이런 부분은 미처 생각하지 못했는데, 초기에 작성했던 훅처럼 반환값이 많아질 경우에는 외부에서 구조를 이해하기 어려울 수 있다는 점을 알게 되었습니다. 그럴 땐 명시적인 타입 선언을 통해 반환 구조를 명확히 드러내주는 것이 더 좋은 방향이라는 점을 알게되었어요!

JeongBin0227 commented on src/pages/order-review/order-coupon/CouponModal.tsx:
```diff
@@ -0,0 +1,134 @@
+  handleApply,
+  isJejuOrRemoteArea,
+}: CouponModalProps) => {
+  const { bestCouponIds } = useOrderCoupons(isJejuOrRemoteArea);
```
useOrderCoupons 이훅이 중복적으로 사용되는거 같은데 props로 주입하는 방식으로 전환 하는것도 좋을거같아요

jeongyou replied:
그렇네요! 굳이 여러 컴포넌트에서 동일한 훅을 반복해서 호출하기보다는, 상위 page에서 useOrderCoupons를 호출한 뒤 bestCouponIds를 props로 전달하도록 수정했습니다

관련 커밋: [refactor: bestCouponIds를 props로 주입받는 방식으로 수정](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/9ada89503cb907e4205a73ec2e617af98865ddd5)

JeongBin0227 commented:
> 1. 상태 관리 방식
> 쿠폰 자동 적용 로직을 위해, 페이지 진입 시점에 totalDiscount와 selectedCouponIds 등의 상태를 설정하고 화면에서 사용하고 있습니다. 하지만 사용자가 쿠폰 선택 모달에 진입한 이후에는 쿠폰을 새롭게 선택할 수 있기 때문에, 모달 내부에서도 별도의 선택 상태를 관리해야 하는 상황입니다.
> 
> 또한 모달 내부에서 선택한 쿠폰에 따라 실시간으로 할인 금액(totalDiscount)도 재계산되어야 하므로, 초기에 설정된 selectedCouponIds와 totalDiscount는 사실상 초기 진입 시점에만 사용되고 이후로는 계속 새로운 값으로 덮어쓰게 됩니다.
> 
> 그리고 모달을 닫고 선택한 쿠폰을 적용할 때, 화면 외부의 상태로 다시 반영하면서 한 번 더 할인 금액 계산이 이루어지기 때문에
> 초기 상태 → 모달 내부 상태 → 최종 적용 흐름 속에서 상태가 여러 군데로 나뉘어 관리되고 있습니다.
> 
> 동기화 자체는 문제 없이 잘 이루어지고 있지만, 상태가 지나치게 분산되고, 일시적으로만 쓰이는 상태들도 많아지다 보니 전체 구조가 복잡해졌다는 느낌을 받고 있습니다. 하나의 상태가 틀어지거나 잘못된 값이 반영되는 경우는 없지만, 유지보수나 로직 이해 측면에서 부담이 커진 것 같습니다.
> 
> ❓이런 상황에서 궁금한 점은 두가지인데요!
> 
> 초기 자동 적용에만 사용되는 상태와, 이후 사용자 선택에 따라 계속 활용되는 상태를 어떻게 명확하게 구분하고 관리하면 좋을지
> SSOT(Single Source of Truth) 원칙을 유지하면서도 자동 적용과 사용자 선택을 유연하게 함께 다룰 수 있는 구조는 어떤 형태가 적절할지
> 케빈과 함께 논의해보고 싶습니다!

->  글을 읽으면서 초기 자동 적용 상태와 사용자 수동 선택 상태의 분리는 개념적으로 다른 책임을 갖는 상태이기 때문에, 명시적으로 구분해서 관리하는 것이 맞다고 생각이 들었어요

초기 자동 적용 상태는 '기준점'으로만 사용
-> bestCouponIds, initialDiscount 등은 단 한 번만 사용되고 그 이후에는 UI에 영향을 주지 않으므로, 상태로 들고 있을 필요 없이 기준값으로만 사용하거나 useRef로만 관리해도 충분합니다.

예: const initialBestCouponIds = useRef(bestCouponIds)

실제 렌더링과 상호작용에 관여하는 상태는 분리된 로컬 상태로 명시
-> selectedCouponIds, modalDiscount 등은 현재 UI의 상태에 직접 관여하므로, 모달 컴포넌트 로컬 상태로 명확하게 구분해서 관리하는 것이 좋을거 같아요.

그럼 SSOT 관점에서의 분리 전략에서는 
페이지 단위: 전체 주문에 대한 쿠폰 상태
모달 단위: 현재 선택된 쿠폰 후보 상태
정리하면, 선택 확정 전까지는 로컬 상태로만 두고, handleApply 시점에만 상위 상태로 반영하면 SSOT 원칙을 해치지 않으면서도 유연한 구조를 유지할 수 있을거 같아요

핵심은 자동 적용 상태는 초기 참조값, 선택 상태는 사용자 인터랙션 상태로 명확히 구분하여 사용 주체에 따라 책임을 분리하는 것입니당!

JeongBin0227 commented:
> 2. 페이지 단위 훅에 대한 고민
> 현재 OrderReviewPage에서는 useOrderInfo, useShipping, useOrderCoupons, useModal, useNavigation 등 관심사 별로 분리된 여러 훅들을 개별적으로 호출하고 있습니다.
> 처음에는 이 훅들을 useOrderReviewPage라는 하나의 훅으로 감싸서 통합적으로 반환할까 고민했지만, 그렇게 할 경우 리턴해야 하는 값이 20개 가까이 되어버려 실제로는 그렇게 구현하지 않았습니다.
> 
> 그렇지만 페이지 내에서 훅이 여러 개 사용되다 보니 코드가 다소 산만해 보인다는 느낌도 있고 어차피 해당 페이지에서만 사용하는 훅들이라면 페이지 단위의 훅 하나로 묶고, 내부에서 관심사 별 훅을 호출한 뒤 필요한 값만 반환하는 구조가 오히려 더 가독성이 좋고 관리하기 쉬운 방향일 수도 있지 않을까? 하는 생각도 들었습니다.
> 
> ❓이런 경우, 하나의 페이지 훅으로 묶는 편이 더 나은 선택인지, 아니면 현재처럼 관심사별 훅들을 페이지에서 개별적으로 사용하는 것이 더 명확한 구조인지에 대해 케빈의 의견이 궁금합니다!

-> useOrderInfo, useShipping, useOrderCoupons 등 관심사 별로 잘 분리된 현재 구조는 재사용성과 독립성 측면에서는 좋은방식인거 같은데, 상황에 따라서 페이지 전용 훅으로 묶는 것을 해야할 수 도 있을거같아요 정리하면,

해당 훅들이 모두 이 페이지 외에서 사용되지 않으면 페이지 훅으로 묶는 게 유리합니다
이런경우 로직상 서로 강하게 연결되어 있어 분리보다는 조합 자체가 의미 있는 경우나 반환해야 할 값이 많아도, 페이지 전용이라면 명시적 인터페이스로 관리 가능한것들일거 같아요

훅들이 다른 페이지나 컴포넌트에서도 재사용될 가능성이 높은 경우는 관심사별 훅을 그대로 사용하는 게 유리합니다.
이런경우는 각 훅의 내부 책임이 명확히 분리되어 있어 단일 책임 원칙을 잘 따르고 있는 경우나 가독성보다 모듈화가 더 중요한 상황이겟죵

정답은 없지만 개인적으로는 해당 훅들이 오직 해당 페이지에서만 의미가 있고 로직이 서로 얽혀 있다면 useOrderReviewPage로 통합해도 무방하다고 생각합니다. 다만 이 경우 useOrderReviewPage의 리턴 타입은 꼭 명시해 주고, 각 책임은 내부 훅에서 분리된 상태로 유지되어야 한다고 생각해요

JeongBin0227 reviewed (changes requested):
안녕하세요 기린~ 
일단 전체적인 코드와 핵심 주제 위주로 코멘트를 드렸고, 코드에 대한 구조를 많이 해주신게 느껴졌는데, 전체적으로 뭔가 정리가 안된 느낌을 많이 받았어요! 코멘트에도 적어놨는데 한번 읽어보면 좋을거 같아요~ 마지막 미션 고생하셨어요~~ 👍

jeongyou committed:
feat: @emotion/react/types/css-prop를 types tsconfig.json의 types에 넣음

jeongyou committed:
refactor: 불필요한 JSX pragma 제거

jeongyou committed:
feat: 선택된 쿠폰들의 할인 금액을 구하는 useCalculateTotalDiscount 훅 구현

jeongyou committed:
feat: 결제 금액을 계산하는 usePayment 훅을 구현

jeongyou committed:
refactor: 페이지에서 직접 currentDiscount, payment를 계산하지 않고 커스텀 훅을 사용하여 계산하도록 수정

jeongyou committed:
refactor: 불필요한 useMemo 제거

jeongyou committed:
refactor: allCouponCombinationIds를 인자로 받지 않고 내부에서 생성해서 사용하도록 변경

jeongyou committed:
refactor: useOrderCoupons 훅에서 allCouponCombinationIds 상태 제거

- 쿠폰 조합 ID 계산을 내부 상태에서 제거하고 책임 분리
- bestCouponIds만으로 필요한 데이터 계산이 가능해 중복 로직 정리
- selectedCouponIds 초기값 설정만 유지하고 관련 useEffect 삭제

jeongyou committed:
feat: 선택된 쿠폰의 상태를 관리하는 useCouponSelection 훅 구현하여 useOrderCoupons에서 로직 분리

jeongyou committed:
refactor: 페이지에서 useCouponSelection 사용하도록 수정

jeongyou committed:
refactor: 할인금액을 직접 계산하지 않고 useCalculateTotalDiscount를 사용하도록 수정

jeongyou committed:
refactor: bestCouponIds를 props로 주입받는 방식으로 수정

jeongyou committed:
fix: 불필요한 totalDiscount props 삭제

jeongyou committed:
style: 불필요한 줄바굼 제거

jeongyou committed:
refactor: 모달 컴포넌트를 역할에 따라 파일별로 분리

jeongyou committed:
refactor: CancelButton 스타일 조건 분기를 객체 매핑 방식으로 리팩토링

jeongyou committed:
refactor: useCallback을 사용하여 handleShowCouponModal 최적화

jeongyou committed:
refactor: 계산 비용이 낮은 useMemo 로직 제거

- usePayment: 단순 산술 연산이므로 useMemo 제거
- useBestCouponCombination: reduce 대상이 매번 새로 생성됨 → memoization 의미 없음
- useCalculateTotalDiscount: 계산 복잡도 낮고 의존성 변화 잦음 → 제거

jeongyou committed:
feat: 선택된 쿠폰들의 할인 금액을 구하는 calculateTotalDiscount 함수 구현

jeongyou committed:
test: 리팩토링된 useBestCouponCombination 훅에 맞춰 테스트 코드 수정

- allCouponCombinationIds 파라미터가 제거되고 내부에서 유틸 호출로 변경됨에 따라,
  getAllCouponCombinationIds 유틸 함수를 mock 처리하여 각 테스트 시나리오 유지

jeongyou committed:
refactor: useCalculateTotalDiscount 훅을 calculateTotalDiscount 로 대체

jeongyou committed:
refactor: 불필요한 console.log 제거

jeongyou committed:
test: useBestCouponCombination 테스트 중복 로직 헬퍼 함수로 리팩토링

jeongyou committed:
test: useAvailableCoupons 테스트 헬퍼 함수로 리팩토링

jeongyou committed:
test: OrderReviewPage 쿠폰 모달 테스트 중복제거

jeongyou committed:
test: CouponDisabled 테스트 유틸 함수 추출 및 중복 제거

jeongyou commented:
안녕하세요 케빈! 잘 지내셨나요? 마지막 미션도 거의 마무리 단계네요…! 😊

케빈의 피드백 덕분에 제가 놓치고 있었던 부분들을 많이 발견할 수 있었고, 기존 코드에서 미처 고려하지 못했던 개념들도 다시 점검해볼 수 있었습니다.

그리고 남겨주신 상태 관리 방식과 페이지 단위 훅에 대한 고민에 대한 답변도 잘 읽어보았습니다. 고민에 대한 충분한 답이 되었고, 또 해당 내용을 참고하여 자동 적용 로직을 리팩터링하는 데 많은 도움이 되었습니다!

**각 리뷰 코멘트에 대한 수정 사항은 반영하였고, 해당 댓글에 답변 남겨두었습니다.
추가적으로 궁금한 사항도 각 코멘트에 질문으로 남겨두었습니다!**

---

그리고 맨 위에 남겨주신 코멘트에 대해 답변을 드리자면

> 코드 중간중간 최적화 (useMemo 등등..) 을 쓰신곳이 있고 아닌곳이 있는데 기준이 궁금합니다!

해당 부분은 로직을 작성하는 과정에서 무한 렌더링 이슈가 발생했을 때, 원인을 추적해보며 “혹시 이 부분 때문인가?” 하는 마음에 useMemo를 여러 곳에 적용하게 되었습니다. 하지만 지금 다시 코드를 살펴보니 대부분의 연산이 단순하고 비용도 크지 않은 부분에 불필요하게 useMemo를 남발했던 것 같습니다...

그래서 리뷰 반영을 하면서 코드 내의 불필요한 useMemo를 모두 제거하였습니다!

관련 커밋: [refactor: 계산 비용이 낮은 useMemo 로직 제거](https://github.com/woowacourse/react-shopping-cart/pull/397/commits/1400eb5a035c398c7c4d5790e9698f6df0b3556c)



> 태스트 코드에 대해서 궁금한데요! 뭔가 정리가 잘되있는거 같으면서도 중구난방한 느낌이 많아서 어떤 기준으로 짜시는지 궁금해요
> 

저는 원래 통합 테스트 위주로 테스트를 작성해왔던 편인데, 이번 미션에서는 솔직하게 시간 관리에 실패하면서 테스트에 충분한 시간을 들이지 못했습니다.

초반에는 페이지 단위로 통합 테스트를 작성했지만, 진행하면서 이번 미션에서 가장 핵심적인 테스트 대상은 쿠폰을 적용하는 훅이 아닐까 하는 생각이 들어서 페이지, 컴포넌트, 훅 등 여러 곳에 테스트를 나눠 작성하다 보니 전체적으로 일관성이 부족한 중구난방한 테스트 구조가 되어버린 것 같습니다.

그리고 아직 테스트 코드에 대한 이해가 깊지 않다 보니, 어떤 부분을 테스트해야 하는지, 내가 중요하다고 생각한 기준으로 테스트를 작성해도 되는지 헷갈리는 지점이 많았습니다.

❓케빈은 테스트 코드를 작성하실 때 어떤 기준이나 우선순위를 가지고 접근하시나요..? 알려주신다면 앞으로 테스트 설계하는 데 큰 도움이 될 것 같습니다..!

> 주석을 자주 쓰시는데 어떤 방향성을 가지고 쓰시는지도 궁금해요

이번 미션의 자동 적용 로직처럼 비교적 복잡하다고 생각되는 부분을 구현할 때, 제가 어떤 의도로 이 코드를 작성했는지 기록하거나, 나중에 작업하거나 리팩터링하면 좋을 부분을 기억하기 위해 주석을 작성하는 경우가 많은 것 같습니다.

또 리팩터링하면서 아직 확신이 없는 코드나 다시 확인이 필요한 부분은 임시로 주석 처리해두는 습관도 있는데 평소에는 리뷰 전에 그런 주석들을 정리하고 요청드리는 편이지만, 이번 미션에서는 시간 부족으로 정리를 다 하지 못한 채 리뷰를 요청드리게 되었습니다.

이번 리뷰에서 주석이 오히려 코드리뷰에 혼란을 줄 수 있다는 점을 느꼈고, 앞으로는 불필요한 주석은 미리 정리하고, 주석을 작성 할 것이라면 목적이 분명하게 전달될 수 있도록 더 신경 써야겠다는 생각이 들었습니다…!

---
케빈도 마지막 미션까지 정말 수고 많으셨습니다! 리뷰 덕분에 많이 배웠습니다! 감사합니다😊🙇‍♀️

JeongBin0227 commented:
> ❓케빈은 테스트 코드를 작성하실 때 어떤 기준이나 우선순위를 가지고 접근하시나요..? 알려주신다면 앞으로 테스트 설계하는 데 큰 도움이 될 것 같습니다..!

-> 일단 기능 기준으로 사용자가 실제로 마주하는 주요 기능들을 먼저 테스트하는거 같아요. 그다음에는 비지니스 로직적으로 중요한 부분, 지금으로 치면 쿠폰 조건, 배송비 계산처럼 중요하거나 분기가 많아 실수 할 수 있는 부분을 주로 작성합니다.

JeongBin0227 reviewed (approved):
안녕하세요~ 기린 수정해주신 코드 모두 확인했습니다. 코드를 단순 반영하지 않고 생각할만한 좋은 주제와 의견을 주셔서 저도 많이 고민하고 배웠던거 같아요! 이번미션은 여기서 머지할게요 미션하느라 고생하셨습니다~ 💯 방학 잘보내세요!

jeongyou commented:
@coderabbitai review

jeongyou commented:
@coderabbitai 리뷰 하고 싶으면 어떻게 해

jeongyou commented:
@coderabbitai 전체적인 리뷰 좀

jeongyou commented:
@coderabbitai 내 코드에서 페인포인트나 안좋은 습관, 개선하면 좋을 부분들이
있어? 있으면 알려줘

jeongyou commented:
@coderabbitai 내 코드에서 불필요한 상태나, useEffect 같은 거 있으면 알려줘
